### ISSUE TITLE
Taking advantage of DokkaConfiguration for CLI runner

### ISSUE THREAD

**Issue by daymxn** [Kotlin/dokka#2579]
Hey folks!

For certain reasons, we have to build our documentation via the CLI runner. In doing so, we take heavy advantage of the JSON configuration. Although, oftentimes our configuration only differs slightly from what a default build may look like, so we end up doing a lot of boilerplate and classpath collection- which otherwise would've been done automatically via dokka. Looking through the source for `DokkaConfiguration` and related classes, it seems you all have already done a bunch of the heavy lifting as far as json serialization, which could be of great use. Is it possible to access the previously configured `DokkaConfiguration` for a given project? Specifically in the context of a Plugin or Task.

tl;dr- If it's possible, how would one access the `DokkaConfiguration` for a project, so that it can be passed to the cli tooling via `toJsonString()`?

**Comment by IgnatBeresnev** [comment]
Hi! Am I right in understanding that you have a number of projects that use Dokka and they configure Dokka in their gradle/maven scripts, but you for some reason cannot use gradle tasks to generate documentation?

So the use case to me looks like the following, correct me if I'm wrong.
1. Have a project with enabled and configured Dokka
2. Dump the configuration to JSON
3. Run the CLI runner with that JSON configuration

If so, there's a function that turns `DokkaConfiguration` into a JSON string, and it's part of public API, so you can call it directly from your gradle/groovy scripts. I haven't checked, but this JSON should be compatible with the CLI runner.

[Source here](https://github.com/Kotlin/dokka/blob/8c218ff4dd5f970233c43845c19299fc74256389/core/src/main/kotlin/configuration.kt#L118)
```kotlin
fun DokkaConfiguration.toJsonString(): String = toJsonString(this)
```

The only question now is how to get project's `DokkaConfiguration`. Unfortunately, [DokkaTask.buildDokkaConfiguration()](https://github.com/Kotlin/dokka/blob/9044761979d08b3b116c9f8416dfb42ae216898c/runners/gradle-plugin/src/main/kotlin/org/jetbrains/dokka/gradle/DokkaTask.kt#L9) is `internal`, so you can't call it directly from the scripts, and I personally cannot find any other way, but maybe you can somehow :) 

For instance, you may try creating a task of your own that extends `DokkaTask` and re-implements [buildDokkaConfiguration()](https://github.com/Kotlin/dokka/blob/9044761979d08b3b116c9f8416dfb42ae216898c/runners/gradle-plugin/src/main/kotlin/org/jetbrains/dokka/gradle/DokkaTask.kt#L9) as a public function. From Java's perspective, `internal` is same as `public`, so maybe creating a Java task will allow you to call the function easily, not sure. I haven't played around with it much, but there might be some wiggle room.

I also believe you could write a Dokka plugin that would use this function to dump json config somewhere (since plugins have access to `DokkaConfiguration`) and do _nothing_ else, but in this case you'd need to be able to call gradle/maven tasks from your project. If that works for you, I can share some more details on how it can be done.

**Comment by daymxn** [comment]
> Hi! Am I right in understanding that you have a number of projects that use Dokka and they configure Dokka in their gradle/maven scripts, but you for some reason cannot use gradle tasks to generate documentation?
> 
> So the use case to me looks like the following, correct me if I'm wrong.
> 
> 1. Have a project with enabled and configured Dokka
> 2. Dump the configuration to JSON
> 3. Run the CLI runner with that JSON configuration
> 
> Since you know about [toJsonString()](https://github.com/Kotlin/dokka/blob/8c218ff4dd5f970233c43845c19299fc74256389/core/src/main/kotlin/configuration.kt#L118), the only question now is how to get project's `DokkaConfiguration`. Unfortunately, [DokkaTask.buildDokkaConfiguration()](https://github.com/Kotlin/dokka/blob/9044761979d08b3b116c9f8416dfb42ae216898c/runners/gradle-plugin/src/main/kotlin/org/jetbrains/dokka/gradle/DokkaTask.kt#L9) is `internal`, so you can't call it directly from the scripts, and I personally cannot find any other way, but maybe you can somehow :)
> 
> For instance, you may try creating a task of your own that extends `DokkaTask` and re-implements [buildDokkaConfiguration()](https://github.com/Kotlin/dokka/blob/9044761979d08b3b116c9f8416dfb42ae216898c/runners/gradle-plugin/src/main/kotlin/org/jetbrains/dokka/gradle/DokkaTask.kt#L9) as a public function. From Java's perspective, `internal` is same as `public`, so maybe creating a Java task will allow you to call the function easily, not sure. I haven't played around with it much, but there might be some wiggle room.
> 
> I also believe you could write a Dokka plugin that would use this function to dump json config somewhere (since plugins have access to `DokkaConfiguration`) and do _nothing_ else, but in this case you'd need to be able to call gradle/maven tasks from your project. If that works for you, I can share some more details on how it can be done.

I appreciate your help!

After some tests, it seems extending `buildDokkaConfiguration` might work, but it ends up being very messy (in Java that is 😢 ). Creating a dokka plugin to dump the json might be a viable solution though- that's a good idea, I'll look more into that. 

Also, as an alternative question, does Dokka officially support extending and utilizing the `impl` methods (`DokkaSourceSetImpl, ExternalDocumentationLinkImpl,  DokkaConfigurationImpl, esc.,`)? Or are they bound to change outside the scope of dokka configuration updates, and are intended to only be used internally? Normally, we'd define our own data class dsls for type safety with the dokka cli- but the impl methods solve that issue for us; unless they're not intended to be prod safe.

**Comment by IgnatBeresnev** [comment]
> does Dokka officially support extending and utilizing the impl methods (DokkaSourceSetImpl, ExternalDocumentationLinkImpl, DokkaConfigurationImpl, esc.,)?

It's hard to tell. Dokka's Plugin API as a whole is in alpha for now and is subject to change, but we'll do our best to mitigate breaking changes, providing time to migrate and clear alternatives.

The only real problem I see is that most of these are data classes, and adding any field on our end would break binary compatibility. This is yet to be resolved

Plugin-wise, have a look at recently updated [Developer guides](https://kotlin.github.io/dokka/1.7.20-SNAPSHOT/developer_guide/introduction/) - it should answer some of the questions on how to write plugins.

You can take our [plugin example](https://github.com/Kotlin/dokka/tree/master/examples/plugin/hide-internal-api) as a starting point and modify it the following way:
```kotlin
class HideInternalApiPlugin : DokkaPlugin() {
    val configurationDumpGeneration by extending {
        CoreExtensions.generation providing ::ConfigurationDumpGeneration override plugin<DokkaBase>().singleGeneration
    }
}

class ConfigurationDumpGeneration(val context: DokkaContext) : Generation {
    override val generationName: String
        get() = "ConfigurationDump"

    override fun Timer.generate() {
        println(context.configuration.toJsonString())
    }
}
```

Publish it to maven local or somewhere, and then use it in your projects as follows:
```kotlin
dokkaPlugin("org.example:hide-internal-api:1.7-SNAPSHOT")
```

Then running any Dokka task would produce what you're looking for:
```
ignat@workstation ~/I/dokka-debug-kts (master) [1]> ./gradlew clean dokkaHtml

> Task :dokkaHtml
Initializing plugins
Dokka is performing: ConfigurationDump
{"moduleName":"dokka-debug-kts","moduleVersion":null,"outputDir":"/home/ignat/IdeaProjects/dokka-debug-kts/build/dokka/html","cacheRoot":null,"offlineMode":false,"sourceSets":[{"displayName":"jvm","sourceSetID":{"scopeId":":dokkaHtml","sourceSetName":"main"},"classpath":["/home/ignat/.m2/repository/org/jetbrains/kotlin/kotlin-stdlib/1.6.21/kotlin-stdlib-1.6.21.jar","/home/ignat/.m2/repository/org/jetbrains/kotlin/kotlin-stdlib-common/1.6.21/kotlin-stdlib-common-1.6.21.jar","/home/ignat/.m2/repository/org/jetbrains/annotations/13.0/annotations-13.0.jar"],"sourceRoots":["/home/ignat/IdeaProjects/dokka-debug-kts/src/main/kotlin","/home/ignat/IdeaProjects/dokka-debug-kts/src/main/java"],"dependentSourceSets":[],"samples":[],"includes":[],"includeNonPublic":false,"reportUndocumented":false,"skipEmptyPackages":true,"skipDeprecated":false,"jdkVersion":8,"sourceLinks":[{"localDirectory":"/home/ignat/IdeaProjects/dokka-debug-kts/src/main/kotlin","remoteUrl":"https://github.com/IgnatBeresnev/dokka-debug-kts/tree/master/src/main/kotlin","remoteLineSuffix":"#L"}],"perPackageOptions":[],"externalDocumentationLinks":[{"url":"https://kotlinlang.org/api/latest/jvm/stdlib/","packageListUrl":"https://kotlinlang.org/api/latest/jvm/stdlib/package-list"},{"url":"https://docs.oracle.com/javase/8/docs/api/","packageListUrl":"https://docs.oracle.com/javase/8/docs/api/package-list"}],"languageVersion":null,"apiVersion":null,"noStdlibLink":false,"noJdkLink":false,"suppressedFiles":[],"analysisPlatform":"jvm","documentedVisibilities":["INTERNAL","PUBLIC"]}],"pluginsClasspath":["/home/ignat/.m2/repository/org/jetbrains/dokka/dokka-analysis/1.7.20-inner-navigastion-4-SNAPSHOT/dokka-analysis-1.7.20-inner-navigastion-4-SNAPSHOT.jar","/home/ignat/.m2/repository/org/example/hide-internal-api/1.7-SNAPSHOT/hide-internal-api-1.7-SNAPSHOT.jar","/home/ignat/.m2/repository/org/jetbrains/dokka/dokka-base/1.7.20-inner-navigastion-4-SNAPSHOT/dokka-base-1.7.20-inner-navigastion-4-SNAPSHOT.jar","/home/ignat/.m2/repository/org/jetbrains/dokka/kotlin-analysis-intellij/1.7.20-inner-navigastion-4-SNAPSHOT/kotlin-analysis-intellij-1.7.20-inner-navigastion-4-SNAPSHOT.jar","/home/ignat/.m2/repository/org/jetbrains/dokka/kotlin-analysis-compiler/1.7.20-inner-navigastion-4-SNAPSHOT/kotlin-analysis-compiler-1.7.20-inner-navigastion-4-SNAPSHOT.jar","/home/ignat/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlinx/kotlinx-html-jvm/0.7.5/f41061b706432f84448a4f1fffa72559dae4d05a/kotlinx-html-jvm-0.7.5.jar","/home/ignat/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlinx/kotlinx-coroutines-core-jvm/1.6.3/b2c2f0c8e99eac632a8db723b365bd59067cf7bd/kotlinx-coroutines-core-jvm-1.6.3.jar","/home/ignat/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-jdk8/1.7.10/d70d7d2c56371f7aa18f32e984e3e2e998fe9081/kotlin-stdlib-jdk8-1.7.10.jar","/home/ignat/.gradle/caches/modules-2/files-2.1/com.fasterxml.jackson.core/jackson-databind/2.12.7/c48e7b68d8cbfec1347e357ee2682e825f861127/jackson-databind-2.12.7.jar","/home/ignat/.gradle/caches/modules-2/files-2.1/com.fasterxml.jackson.core/jackson-annotations/2.12.7/2042461b754cd65ab2dd74a9f19f442b54625f19/jackson-annotations-2.12.7.jar","/home/ignat/.gradle/caches/modules-2/files-2.1/com.fasterxml.jackson.core/jackson-core/2.12.7/4669a54b799c105572aa8de2a1ae0fe64a17745/jackson-core-2.12.7.jar","/home/ignat/.gradle/caches/modules-2/files-2.1/com.fasterxml.jackson.module/jackson-module-kotlin/2.12.7/9dd38ab8f194300be73a0f1e13701b9b1e012337/jackson-module-kotlin-2.12.7.jar","/home/ignat/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-reflect/1.7.10/165e600dfea6185cf5efa700756294cc4904dbeb/kotlin-reflect-1.7.10.jar","/home/ignat/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-jdk7/1.7.10/1ef73fee66f45d52c67e2aca12fd945dbe0659bf/kotlin-stdlib-jdk7-1.7.10.jar","/home/ignat/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib/1.7.10/d2abf9e77736acc4450dc4a3f707fa2c10f5099d/kotlin-stdlib-1.7.10.jar","/home/ignat/.gradle/caches/modules-2/files-2.1/org.jsoup/jsoup/1.14.3/c43a81e18e6d0eb71951aa031d55d5c293c531a6/jsoup-1.14.3.jar","/home/ignat/.gradle/caches/modules-2/files-2.1/org.freemarker/freemarker/2.3.31/cd4fc0942b4a8bdb19f3b669aa42136fb54feb55/freemarker-2.3.31.jar","/home/ignat/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-common/1.7.10/bac80c520d0a9e3f3673bc2658c6ed02ef45a76a/kotlin-stdlib-common-1.7.10.jar","/home/ignat/.m2/repository/org/jetbrains/annotations/13.0/annotations-13.0.jar"],"pluginsConfiguration":[],"modules":[],"failOnWarning":false,"delayTemplateSubstitution":false,"suppressObviousFunctions":true,"includes":[],"suppressInheritedMembers":false,"finalizeCoroutines":true}
```

You can of course change all the names, artifactId and maybe create a separate task for it, like `./gradlew dumpDokkaConfig` or something.

Hopefully that answers the question, so I'm closing it for now. If you get stuck or something is unclear, get in touch, we'll do our best to help :)
